src/controller/certificate.rs

```
@@ -0,0 +1,588 @@
+                    match result {
+                        Ok(authz) => {
+                            let identifier = authz.identifier().to_string();
+                            match authz.status {
+                                AuthorizationStatus::Valid => {
+                                    info!("Authorization for {} is already valid", identifier);
+                                    continue;
+                                }
+                                AuthorizationStatus::Pending => {
+                                    if authz
+                                        .challenges
+                                        .iter()
+                                        .any(|c| c.r#type == ChallengeType::Http01)
+                                    {
+                                        needs_challenge = true;
+                                        info!("Found HTTP-01 challenge for {}", identifier);
+                                    } else {
+                                        error!("No supported challenge type for {}", identifier);
+                                        status.state = CertificateState::Failed;
+                                        status.last_failure_reason = Some(format!(
+                                            "No supported challenge type for {}",
+                                            identifier
+                                        ));
+                                        return Ok(ReconcileNext::After(Duration::from_secs(60)));
+                                    }
+                                }
+                                _ => {
+                                    error!(
+                                        "Authorization for {} in unexpected state: {:?}",
+                                        identifier, authz.status
+                                    );
+                                    status.state = CertificateState::Failed;
```

you could just return the error from reconciliation. it will reach `handle_error` in the controller where you can set the state to failed, set the last fail reason to the message from the error and add the ReconcileNext::After call. this is why `handle_error` exists. to dedupe this kind of error handling.